$100$ Hz と $200$ Hz をそれぞれメル周波数に変換し，それらの差 $M_1$ を計算しなさい．
また， $4000$ Hz と $4100$ Hz をそれぞれメル周波数に変換し，それらの差 $M_2$ を計算しなさい．
$4000$ Hz よりメル尺度で $M_1$ だけ大きい周波数を計算し，その周波数を持つ正弦波を 3 と同様に作り，保存しなさい．
$4000$ Hz の音とその音の高さの差は，$100$ Hz の音と $200$ Hz の音の高さの差と同じように感じられるか？

In [1]:
from pathlib import Path
import numpy as np
import soundfile as sf
from IPython.display import Audio, display

def hz_to_mel(f):
    return 2595 * np.log10(1 + f / 700)

def mel_to_hz(m):
    return 700 * (10 ** (m / 2595) - 1)

freqs = [100, 200, 4000, 4100]
mels = {f: hz_to_mel(f) for f in freqs}

M1 = mels[200] - mels[100]
M2 = mels[4100] - mels[4000]

target_mel = mels[4000] + M1
target_hz = mel_to_hz(target_mel)

print("周波数 [Hz]\tメル周波数 [mel]")
print("-" * 34)
for f in freqs:
    print(f"{f:>8}\t{mels[f]:>12.3f}")

print()
print(f"M1 = mel(200) - mel(100) = {M1:.3f} mel")
print(f"M2 = mel(4100) - mel(4000) = {M2:.3f} mel")
print(f"mel(4000) + M1 に対応する周波数 = {target_hz:.3f} Hz")


周波数 [Hz]	メル周波数 [mel]
----------------------------------
     100	     150.489
     200	     283.230
    4000	    2146.065
    4100	    2169.792

M1 = mel(200) - mel(100) = 132.741 mel
M2 = mel(4100) - mel(4000) = 23.727 mel
mel(4000) + M1 に対応する周波数 = 4587.500 Hz


計算結果は次の通り．

- $M_1 \simeq 132.741\,\mathrm{mel}$
- $M_2 \simeq 23.727\,\mathrm{mel}$
- $4000\,\mathrm{Hz}$ よりメル尺度で $M_1$ だけ大きい周波数は，約 $4587.5\,\mathrm{Hz}$

$M_1 > M_2$ より，同じ $100\,\mathrm{Hz}$ 差でも低周波側の方がメル尺度上の差が大きい．
$4000\,\mathrm{Hz}$ と $4587.5\,\mathrm{Hz}$ のメル差は，
$100\,\mathrm{Hz}$ と $200\,\mathrm{Hz}$ のメル差と同じになるように設計された値．

ただし，実際の聴感は個人差，再生環境，音量に依存するため，完全に同じように感じるとは限らない．


In [2]:
output_dir = Path("outputs/audio")
output_dir.mkdir(parents=True, exist_ok=True)

fs = 16000
duration = 5.0
amplitude = 0.5

t = np.arange(int(fs * duration)) / fs
x_4000 = amplitude * np.sin(2 * np.pi * 4000 * t)
x_target = amplitude * np.sin(2 * np.pi * target_hz * t)

path_4000 = output_dir / "04_sine_4000hz.wav"
path_target = output_dir / "04_sine_4587_5hz_mel_equal.wav"

sf.write(path_4000, x_4000, fs)
sf.write(path_target, x_target, fs)

print(f"保存先: {path_4000}")
print(f"保存先: {path_target}")

print("4000 Hz")
display(Audio(x_4000, rate=fs))

print(f"{target_hz:.1f} Hz")
display(Audio(x_target, rate=fs))


保存先: outputs/audio/04_sine_4000hz.wav
保存先: outputs/audio/04_sine_4587_5hz_mel_equal.wav
4000 Hz


4587.5 Hz
